pkg/hbone/doubledialer.go

```diff
@@ -0,0 +1,269 @@
+		ids := resp.TLS.PeerCertificates[0].SerialNumber
+		remoteID = ids.String()
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("round trip failed: %v", resp.Status)
+	}
+	log.WithLabels("host", r.Host, "remote", remoteID).Info("Outer CONNECT established")
+
+	log.Infof("initiate inner CONNECT to %v via inner tunnel", r.Host)
+
+	ipr, ipw := io.Pipe()
+	innerReq, err := http.NewRequest(http.MethodConnect, url, ipr)
+	if err != nil {
+		return fmt.Errorf("new inner request: %v", err)
+	}
+
+	pc := &pipeConn{
+		name:   "inner CONNECT",
+		writer: pw,
+		reader: resp.Body,
+	}
+	innerReq.Host = address
+	var innerTransport *http2.Transport
+	if d.innerTLSConfig != nil {
+		log.Infof("using TLS on inner connection")
+		innerTransport = &http2.Transport{
+			TLSClientConfig: d.innerTLSConfig,
+			DialTLSContext: func(ctx context.Context, network, addr string, tlsCfg *tls.Config) (net.Conn, error) {
+				tlsCfg.ServerName = r.Host
+				// Upgrade the raw connection to a TLS connection.
+				c := tls.Client(pc, tlsCfg)
+				err := c.Handshake()
+				if err != nil {
+					pc.Close()
+					return nil, err
+				}
+				return c, nil
+			},
+		}
+	} else {
+		innerTransport = &http2.Transport{
+			// For h2c
+			AllowHTTP: true,
+			DialTLSContext: func(ctx context.Context, network, addr string, tlsCfg *tls.Config) (net.Conn, error) {
+				return pc, nil
+			},
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	// NOTE: We don't close the connection here because that will affect the outer connection from the HBONE client.
+	// Instead, we close the conn further down once we've finished proxying.
+	go func() {
+		// handle upstream (inner hbone server) <-- downstream (app)
+		copyBuffered(ipw, conn, log.WithLabels("name", "body to pipe"))
+		wg.Done()
+	}()
+
+	innerResp, innerErr := innerTransport.RoundTrip(innerReq)
+	if innerErr != nil {
+		return fmt.Errorf("inner round trip: %v", innerErr)
+	}
+	log.Info("inner round trip complete")
+
+	if innerResp.TLS != nil {
+		log.Infof("inner TLS does exist")
+		if len(innerResp.TLS.PeerCertificates) > 0 {
+			ids := innerResp.TLS.PeerCertificates[0].SerialNumber
+			innerID = ids.String()
+		}
+	}
+	if innerResp.StatusCode != http.StatusOK {
+		return fmt.Errorf("round trip failed: %v", innerResp.Status)
+	}
+	log.WithLabels("host", r.Host, "remote", innerID).Info("Inner CONNECT established")
+
+	go func() {
+		defer conn.Close()
+		defer resp.Body.Close()
+
+		wg.Add(1)
+		go func() {
+			// handle upstream (hbone server) --> downstream (app)
+			copyBuffered(conn, innerResp.Body, log.WithLabels("name", "body to conn"))
+			wg.Done()
+		}()
+
+		wg.Wait()
+		log.Infof("stream closed in %v", time.Since(t0))
+	}()
+
+	return nil
+}
+
+type pipeAddr struct {
+	name string
+}
+
+func (p *pipeAddr) Network() string {
+	return "pipe"
+}
+
+func (p *pipeAddr) String() string {
+	return "pipe " + p.name
+}
+
+// implements net.Conn
+type pipeConn struct {
```

Comment:
My head is spinning a bit with all the layers and pipes and connections but I think when we did the single hbone we avoided this by utilizing `net.Pipe`. You probably just need to make sure you have both copyBuffered directions (like in the single hbone) instead of the one we have here.

not relevant if we are able to collapse this into a single "HBONE" logic shared for single and double, though (triple??)

Reply:
The reason why `net.Pipe` doesn't work is because you can't compose them. You can't tell `net.Pipe` that the reads are coming from an existing `io.Reader`; same for writes. That's what we need to make double hbone work

Reply:
I was suggesting to `io.copy` between the reader and the net.Pipe. Then you can just compose them.

Then we have two cases that are really one code.
1. HBONE (for inner HBONE) over a `io.ReadWriteCloser` (backed by `net.TCPConn`)
2. HBONE (for double HBONE) over a `io.ReadWriteCloser` (backend by a `net.Pipe`)

Reply:
I gave this a try, and I think the semantics of net.Pipe aren't what we want. With `net.Pipe` reads on one end are matched with writes on the other. I don't think that's actually what we want; when I test this out, I get tls handshake errors on the inner connection. With pipeConn, what we're saying is, _when_ you read, read from this `io.Reader` and _when_ you write, write to this `io.Writer`.

Reply:
hmmm. regardless the reader/writer type I think we could still make 1 common hbone logic that is abstracted over the underlying connection rather than 2 split paths. Not a blocker though, just a suggestion